File: BeehiveMac/Sources/BeehiveKit/Record.swift
import Foundation

/// Metadata for one encoded song (scalars + definitions). Codable for the ledger.
public struct HelixMeta: Codable {
    public var engineVersion = "0.1.0-swift"
    public var songId: String
    public var sourcePath: String
    public var contentHash: String
    public var createdAt: String
    public var sr: Double
    public var nSamples: Int
    public var durationS: Double
    public var nFFT: Int
    public var hop: Int
    public var win: Int
    public var nFrames: Int
    public var bpm: Double
    public var bpmConfidence: Double?
    public var bpmSource: String
    public var B: Int
    public var omegaRadPerS: Double
    public var omegaRadPerFrame: Double
    public var r0: Double
    public var fRef: Double
    public var sigma: Double
    public var floorDB: Double
    public var wAmp: Double
    public var wInt: Double
    public var smoothFrames: Int

    public init(engineVersion: String = "0.1.0-swift", songId: String, sourcePath: String,
                contentHash: String, createdAt: String, sr: Double, nSamples: Int,
                durationS: Double, nFFT: Int, hop: Int, win: Int, nFrames: Int, bpm: Double,
                bpmConfidence: Double?, bpmSource: String, B: Int, omegaRadPerS: Double,
                omegaRadPerFrame: Double, r0: Double, fRef: Double, sigma: Double,
                floorDB: Double, wAmp: Double, wInt: Double, smoothFrames: Int) {
        self.engineVersion = engineVersion; self.songId = songId; self.sourcePath = sourcePath
        self.contentHash = contentHash; self.createdAt = createdAt; self.sr = sr
        self.nSamples = nSamples; self.durationS = durationS; self.nFFT = nFFT; self.hop = hop
        self.win = win; self.nFrames = nFrames; self.bpm = bpm; self.bpmConfidence = bpmConfidence
        self.bpmSource = bpmSource; self.B = B; self.omegaRadPerS = omegaRadPerS
        self.omegaRadPerFrame = omegaRadPerFrame; self.r0 = r0; self.fRef = fRef
        self.sigma = sigma; self.floorDB = floorDB; self.wAmp = wAmp; self.wInt = wInt
        self.smoothFrames = smoothFrames
    }
}

/// The persistent, on-standby encoding of one song. Only the independent state
/// (A, I, chroma, engagement) is stored; everything geometric is recomputed by
/// Engine.deriveDynamics on load. Port of beehive/record.py.
public struct HelixRecord {
    public let meta: HelixMeta
    // stored state
    public let A: [Float]
    public let I: [Float]
    public let chroma: [[Float]]          // nFrames x 12
    public var engagement: [Float]
    // derived
    public let t: [Double]
    public let dA: [Double]
    public let dI: [Double]
    public let fMag: [Double]
    public let h: [Double]
    public let alpha: [Double]
    public let x: [Double]
    public let y: [Double]
    public let z: [Double]
    public let kappa: [Double]
    public let tau: [Double]
    public let hue: [Float]
    public let sat: [Float]
    public let val: [Float]

    public var nFrames: Int { A.count }

    /// What this compact representation costs vs. the dense STFT it replaces.
    public func compressionReport() -> (stored: Int, stft: Int, ratioState: Int, ratioStored: Int) {
        let n = nFrames, bins = meta.nFFT / 2 + 1
        let state = 2 * n * 4
        let stored = (2 + 12) * n * 4
        let stft = bins * n * 4
        return (stored, stft, stft / max(state, 1), stft / max(stored, 1))
    }
}

/// Codable wrapper for persisting only the stored state.
struct StoredState: Codable {
    let meta: HelixMeta
    let A: [Float]
    let I: [Float]
    let chroma: [[Float]]
    let engagement: [Float]
}

public extension HelixRecord {
    func save(to url: URL) throws {
        let s = StoredState(meta: meta, A: A, I: I, chroma: chroma, engagement: engagement)
        let data = try JSONEncoder().encode(s)
        try data.write(to: url)
    }

    static func load(from url: URL) throws -> HelixRecord {
        let s = try JSONDecoder().decode(StoredState.self, from: Data(contentsOf: url))
        return Engine.assemble(meta: s.meta, A: s.A, I: s.I, chroma: s.chroma, engagement: s.engagement)
    }
}

File: beehive/record.py
"""HelixRecord: the persistent, on-standby encoding of one song.

This is the artifact the project is really after -- a song "analysed once" and
kept, so it can later be run against (taste profile) and have a listener
engagement layer accumulated over it. It stores only the *compact per-frame
state* and the geometry derived from it, never the dense STFT it is meant to
replace.

Layout per record (n = n_frames):
  state      A, I            -- amplitude (dB), interval (semitones)
  force      dA, dI, F_mag   -- the force vector and its magnitude ||F|| (energy spent)
  geometry   h, alpha, x,y,z -- climb, angle, helix position; plus kappa, tau
  colour     chroma(n,12), hue, sat, val   -- notes as lights, dictated by timbre
  engagement engagement(n)   -- RESERVED: the "most rewatched, but for music" layer
                               (zeros until real playback capture exists)

On disk a record is two files: <stem>.npz (arrays) and <stem>.json (metadata).
"""

import json
from dataclasses import dataclass, field

import numpy as np

# Only the *independent* state is stored on disk; everything else is a function
# of these plus the scalar meta and is recomputed by encode.derive_dynamics on
# load. (A dense STFT is what we refuse to keep -- this is the whole compactness
# claim.)  A=loudness, I=interval, chroma=timbre, engagement=listener layer.
STORED_FIELDS = ["A", "I", "chroma", "engagement"]


@dataclass
class HelixRecord:
    meta: dict
    t: np.ndarray
    A: np.ndarray
    I: np.ndarray
    dA: np.ndarray
    dI: np.ndarray
    F_mag: np.ndarray
    h: np.ndarray
    alpha: np.ndarray
    x: np.ndarray
    y: np.ndarray
    z: np.ndarray
    kappa: np.ndarray
    tau: np.ndarray
    chroma: np.ndarray
    hue: np.ndarray
    sat: np.ndarray
    val: np.ndarray
    engagement: np.ndarray = field(default=None)

    @property
    def n_frames(self):
        return int(self.A.shape[0])

    def save(self, stem):
        """Write <stem>.npz (the stored state only) and <stem>.json (metadata)."""
        arrays = {k: np.asarray(getattr(self, k), dtype=np.float32) for k in STORED_FIELDS}
        np.savez_compressed(f"{stem}.npz", **arrays)
        with open(f"{stem}.json", "w", encoding="utf-8") as fh:
            json.dump(self.meta, fh, indent=2)
        return f"{stem}.npz"

    @classmethod
    def load(cls, stem):
        from .encode import derive_dynamics    # local import avoids a cycle

        stem = stem[:-4] if stem.endswith(".npz") else stem
        with np.load(f"{stem}.npz") as data:
            stored = {k: data[k] for k in STORED_FIELDS}
        with open(f"{stem}.json", encoding="utf-8") as fh:
            meta = json.load(fh)
        der = derive_dynamics(stored["A"], stored["I"], stored["chroma"], meta)
        return cls(meta=meta, **stored, **der)

    def compression_report(self):
        """Logical bytes of this representation vs. the dense STFT it replaces.

        Honest accounting (uncompressed float32):
          state  = 2 floats/frame (A, I) -- the helix geometry substrate;
          stored = 14 floats/frame (A, I, chroma[12]) -- what the .npz holds
                   (engagement is reserved zeros and compresses away);
          stft   = n_bins floats/frame -- the dense magnitude spectrogram.

        This counts columns, not adaptive offload: ||F|| says *where* the music
        moves, but every frame is currently stored at the same density. For the
        on-disk (compressed) size, stat the .npz directly.
        """
        n = self.n_frames
        n_bins = int(self.meta["n_fft"] // 2 + 1)
        state = 2 * n * 4                     # A, I
        stored = (2 + 12) * n * 4            # A, I, chroma  (engagement ~ free)
        stft = n_bins * n * 4
        return {
            "n_frames": n,
            "n_bins": n_bins,
            "state_bytes": state,
            "stored_bytes": stored,
            "stft_bytes": stft,
            "ratio_state_vs_stft": stft / max(state, 1),
            "ratio_stored_vs_stft": stft / max(stored, 1),
        }
